# SEmRNA-seq-nf

A Nextflow pipeline used for quantification and quanlity control for RNA-seq data.

## Usage

Clone this pipeline use command below:
```
git clone git@github.com:AndersenLab/SEmRNA-seq-nf.git
```
and `cd SEmRNA-seq-nf`

For local run, if you installed [docker](https://www.docker.com), you can run the pipeline with conmmand:
```
nextflow run main.nf -with-docker -profile local
```
Else:
```
nextflow run main.nf -profile local
```

To run this pipeline on [quest](http://www.it.northwestern.edu/research/user-services/quest/index.html), please use command:
```
nextflow run main.nf -profile quest
```

### `main.nf` and `test.nf`

The `main.nf` is the default setting for this pipeline that using a sample sheet as input.
The sample sheet should strictly follow the format showed in `test.tsv`.
Each column represent `strain_name` `sample_name` `ID` `sample_order` `absolute directory` `sequence_folder` `sub_folder` `unique_lable`
Note that `sample_name` `ID` `absolute directory` `uniq_lable` should be unique for each sequence.

`test.nf` is an option for those who do not want to creat a sample sheet. `test.nf` reads the path for raw sequences directly.

They have no effect on results. 

## Pipeliine parameters

* --fqs

We use a sample sheet as the input of sequences here, see the example in `test.tsv`.

* --transcriptome

Transcriptome used for reads mapping. 
This pipeline use [alignment-independent](https://cgatoxford.wordpress.com/2016/08/17/why-you-should-stop-using-featurecounts-htseq-or-cufflinks2-and-start-using-kallisto-salmon-or-sailfish/) methods for quantification.
Thus, we don't need map to genome.

* --output

Used to specify the output directory

* --fragment_len

This is only used for single-end reads by kallisto. Pair-end reads do not need to set this parameter. see the [kallisto document](https://pachterlab.github.io/kallisto/manual) for details.
* --fragment_sd

This is only used for single-end reads by kallisto. Pair-end reads do not need to set this parameter. see the [kallisto document](https://pachterlab.github.io/kallisto/manual) for details.
* --bootstrap

Number of bootstrap samples used by kallisto.

* --experiment

This is an important params for pipeline, see the reference format in `experiment_info.txt`. The style should be strictly followed.
```
--experiment <experiment_info>
```

* --email

Add your email with command
```
--emial < ...@...>
```

When the pipeline completed or broken, you can receive the notification from your email.

## Output

This pipeline will generate a `report.html` and `trace.txt` in your working directory.

Below are major results under `results/`.
```
├── log.txt                 # Running information
├── multiqc_report.html     # A conbined quanlity control results
├── sleuth_object.so        # Quantification results, open it in R with sleuth package
└── transcriptome.index     # Index of transcriptome generated by kallisto
```

## Dependencies

* [Nextflow](git@github.com:AndersenLab/SEmRNA-seq-nf.git)

* [Kallisto](https://pachterlab.github.io/kallisto/)

* [Salmon](https://combine-lab.github.io/salmon/)

* Fastqc

* [Multiqc](http://multiqc.info)

* [Sleuth](https://pachterlab.github.io/sleuth/)


